### PR TITLE
[5.0.0] Big Breaking Android Refactoring

### PR DESCRIPTION
**Pull request by ngoryachev**

### CODE REVIEW

File: android/src/main/java/com/gettipsi/stripe/PayFlow.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+public abstract class PayFlow {
+
+  public static final String NO_CURRENT_ACTIVITY_MSG = "Cannot start process with no current activity";
+  public static final String PURCHASE_CANCELLED_MSG = "Purchase was cancelled";
+  public static final String PURCHASE_LOAD_MASKED_WALLET_ERROR_MSG = "Purchase masked wallet error";
+  public static final String PURCHASE_LOAD_FULL_WALLET_ERROR_MSG = "Purchase full wallet error";
+  public static final String ANDROID_PAY_UNAVAILABLE_ERROR_MSG = "Android Pay is unavailable";
+  public static final String MAKING_IS_READY_TO_PAY_CALL_ERROR_MSG = "Error making isReadyToPay call";
+  public static final String JSON_PARSING_ERROR_MSG = "Failed to create token from JSON string.";
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/PayFlow.java]: the only dot in this list

File: src/Stripe.android.js [hunk 2]
```diff
@@ -21,6 +21,8 @@ class Stripe {
     StripeModule.deviceSupportsAndroidPay()
   )
 
+  setAndroidPayMode = StripeModule.setAndroidPayMode
```
Comment (isnifer) [inline src/Stripe.android.js]:
```js
setAndroidPayMode = mode => (
  StripeModule.setAndroidPayMode(mode)
)
```

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java [hunk 3]
```diff
@@ -160,82 +91,68 @@ public String getName() {
...
-          .build())
-        .build();
-      googleApiClient.connect();
-    } else {
-      promise.reject(TAG, "Unknown error");
-    }
+  private static int androidPayModeToEnvironment(@NonNull String androidPayMode) {
+    ArgCheck.notEmptyString(androidPayMode);
+    int environment = ANDROID_PAY_MODE_TEST.equals(androidPayMode.toLowerCase()) ? WalletConstants.ENVIRONMENT_TEST : WalletConstants.ENVIRONMENT_PRODUCTION;
+    return environment;
```
Comment (paynd) [inline android/src/main/java/com/gettipsi/stripe/StripeModule.java]: return result without var creation

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java [hunk 4]
```diff
@@ -160,82 +91,68 @@ public String getName() {
...
+  }
+
+  @ReactMethod
+  public void setAndroidPayMode(String androidPayMode) {
```
Comment (paynd) [inline android/src/main/java/com/gettipsi/stripe/StripeModule.java]: Does we need to reassign pay mode?

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java [hunk 5]
```diff
@@ -351,7 +272,7 @@ public void createSourceWithParams(final ReadableMap options, final Promise prom
 
     }
 
-    stripe.createSource(sourceParams, new SourceCallback() {
+    mStripe.createSource(sourceParams, new SourceCallback() {
```
Comment (paynd) [inline android/src/main/java/com/gettipsi/stripe/StripeModule.java]: can `sourceParams` be null if here?

File: android/src/main/java/com/gettipsi/stripe/PayFlow.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  abstract void deviceSupportsAndroidPay(final Promise promise);
+
+  abstract boolean onActivityResult(Activity activity, int requestCode, int resultCode, Intent data);
+
+  public static boolean isPlayServicesAvailable(@NonNull Activity activity) {
+    ArgCheck.nonNull(activity);
+
+    GoogleApiAvailability googleAPI = GoogleApiAvailability.getInstance();
+    int result = googleAPI.isGooglePlayServicesAvailable(activity);
+    if (result != ConnectionResult.SUCCESS) {
```
Comment (paynd) [inline android/src/main/java/com/gettipsi/stripe/PayFlow.java]: Better `return result == ConnectionResult.SUCCESS;`

File: android/src/main/java/com/gettipsi/stripe/PayFlow.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      return false;
+    }
+    return true;
+  }
+
+  private static void log(String TAG, String msg) {
```
Comment (paynd) [inline android/src/main/java/com/gettipsi/stripe/PayFlow.java]: We dont use it anywhere? And it's `private` :)

File: android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        .addAllowedPaymentMethod(WalletConstants.PAYMENT_METHOD_CARD)
+        .addAllowedPaymentMethod(WalletConstants.PAYMENT_METHOD_TOKENIZED_CARD)
+        .build();
+    mPaymentsClient = createPaymentsClient(activity);
+    Task<Boolean> task = mPaymentsClient.isReadyToPay(request);
+    task.addOnCompleteListener(
+      new OnCompleteListener<Boolean>() {
+        public void onComplete(Task<Boolean> task) {
+          try {
+            boolean result = task.getResult(ApiException.class);
```
Comment (paynd) [inline android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java]: Can we just:
`promise.resolve(task.getResult(ApiException.class))`
?

Comment (ngoryachev) [reply android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java]: @paynd I left it as is for simpler logging insertion later. Please let me know if using cheap stack-var is critical for you.

File: android/src/main/java/com/gettipsi/stripe/PayFlow.java (hunk 7 above)
Comment (ngoryachev) [reply android/src/main/java/com/gettipsi/stripe/PayFlow.java]: made it protected and left as possibility for future.

File: android/src/main/java/com/gettipsi/stripe/PayFlow.java (hunk 6 above)
Comment (ngoryachev) [reply android/src/main/java/com/gettipsi/stripe/PayFlow.java]: done

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java (hunk 5 above)
Comment (ngoryachev) [reply android/src/main/java/com/gettipsi/stripe/StripeModule.java]: Nice catch, made an assertion.

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java (hunk 4 above)
Comment (ngoryachev) [reply android/src/main/java/com/gettipsi/stripe/StripeModule.java]: Nice catch, removed it unused method.

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java (hunk 3 above)
Comment (ngoryachev) [reply android/src/main/java/com/gettipsi/stripe/StripeModule.java]: done

File: android/src/main/java/com/gettipsi/stripe/PayFlow.java (hunk 1 above)
Comment (ngoryachev) [reply android/src/main/java/com/gettipsi/stripe/PayFlow.java]: done. Eagle eye!

File: src/Stripe.android.js (hunk 2 above)
Comment (ngoryachev) [reply src/Stripe.android.js]: it's unused now so i removed that method.

File: android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+              WalletConstants.CARD_NETWORK_AMEX,
+              WalletConstants.CARD_NETWORK_DISCOVER,
+              WalletConstants.CARD_NETWORK_VISA,
+              WalletConstants.CARD_NETWORK_MASTERCARD))
+          .build())
+      .addAllowedPaymentMethod(WalletConstants.PAYMENT_METHOD_CARD)
+      .addAllowedPaymentMethod(WalletConstants.PAYMENT_METHOD_TOKENIZED_CARD)
+      .setShippingAddressRequired(shippingAddressRequired);
+
+    if(countryCodes.size() > 0) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+      LOAD_PAYMENT_DATA_REQUEST_CODE);
+  }
+
+  @Override
+  public void paymentRequestWithAndroidPay(@NonNull ReadableMap payParams, @NonNull Promise promise) {
+    ArgCheck.nonNull(payParams);
+    ArgCheck.nonNull(promise);
+
+    Activity activity = activityProvider.call();
+    if(activity == null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    this.payPromise = promise;
+    startPaymentRequest(activity, createPaymentDataRequest(payParams));
+  }
+
+  @Override
+  public void deviceSupportsAndroidPay(@NonNull Promise promise) {
+    Activity activity = activityProvider.call();
+    if(activity == null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    if (!isPlayServicesAvailable(activity)) {
+      promise.reject(TAG, PLAY_SERVICES_ARE_NOT_AVAILABLE_MSG);
+      return;
+    }
+
+    isReadyToPay(activity, promise);
+  }
+
+  public boolean onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
+    if(payPromise == null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+  private ReadableMap mAndroidPayParams;
+
+  public ObsoleteApiPayFlowImpl(@NonNull Fun0<Activity> activityProvider) {
+    super(activityProvider);
+  }
+
+  @Override
+  public void deviceSupportsAndroidPay(final Promise promise) {
+    Activity activity = activityProvider.call();
+    if(activity == null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+      promise.reject(TAG, PLAY_SERVICES_ARE_NOT_AVAILABLE_MSG);
+      return;
+    }
+
+    connectAndCheckIsWalletReadyToPay(promise);
+  }
+
+  @Override
+  public boolean onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
+    if(mPayPromise == null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+      if (token == null) {
+        // Log the error and notify Stripe help
+        mPayPromise.reject(TAG, JSON_PARSING_ERROR_MSG);
+      } else {
+        mPayPromise.resolve(putExtraToTokenMap(
+          convertTokenToWritableMap(token),
+          fullWallet.getBuyerBillingAddress(),
+          fullWallet.getBuyerShippingAddress()));
+      }
+    } else if(resultCode == Activity.RESULT_CANCELED) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+      }},
+      new Action<Bundle>() { public void call(Bundle bundle) {
+        isWalletReadyToPay(mGoogleApiClient, promise);
+      }});
+    mGoogleApiClient.connect();
+  }
+
+  private void connectAndCheckIsWalletReadyToPayAndLoadMaskedWallet(final ReadableMap payParams) {
+    final Activity activity = activityProvider.call();
+    if(activity == null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+        disconnect(mGoogleApiClient);
+      }},
+      new Action<Bundle>() { public void call(Bundle bundle) {
+        checkIsWalletReadyToPayAndLoadMaskedWallet(mGoogleApiClient, mPayPromise, payParams, activity);
+      }});
+    mGoogleApiClient.connect();
+  }
+
+  private static void disconnect(GoogleApiClient googleApiClient) {
+    if(googleApiClient != null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java [hunk 18]
```diff
@@ -0,0 +1,319 @@
...
+      mPayPromise.reject(TAG, PURCHASE_LOAD_MASKED_WALLET_ERROR_MSG);
+      disconnect(mGoogleApiClient);
+    }
+  }
+
+  private static void isWalletReadyToPay(final GoogleApiClient googleApiClient, final Promise promise) {
+    isWalletReadyToPay(
+      googleApiClient,
+      new ResultCallback<BooleanResult>() { public void onResult(@NonNull BooleanResult result) {
+        if(result.getStatus().isSuccess()) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/ObsoleteApiPayFlowImpl.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/PayFlow.java [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+  private int environment; // invalid value by default
+
+
+  public PayFlow(@NonNull Fun0<Activity> activityProvider) {
+    ArgCheck.nonNull(activityProvider);
+    this.activityProvider = activityProvider;
+  }
+
+  public static PayFlow create(Fun0<Activity> activityProvider) {
+    if(!useObsoleteFlow) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/PayFlow.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/ArgCheck.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
+package com.gettipsi.stripe.util;
+
+import android.text.TextUtils;
+
+public abstract class ArgCheck {
+
+    public static <T> T nonNull(T t) {
+        if(t == null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/ArgCheck.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/ArgCheck.java [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+            throw new NullPointerException();
+        }
+
+        return t;
+    }
+
+    public static String notEmptyString(String string) {
+        if(TextUtils.isEmpty(string)) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/ArgCheck.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/ArgCheck.java [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public static String isDouble(String string) {
+        String digital = notEmptyString(string);
+        Double.parseDouble(string);
+        return digital;
+    }
+
+    public static void isTrue(boolean shouldBeTrue) {
+        if(!shouldBeTrue) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/ArgCheck.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/Action.java [hunk 23]
```diff
@@ -0,0 +1,5 @@
+package com.gettipsi.stripe.util;
+
+public interface Action<T> {
+    void call(T t);
+}
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Action.java]: There are too many files with `2 spaces` as indent

File: android/src/main/java/com/gettipsi/stripe/util/ArgCheck.java [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+            throw new IllegalArgumentException();
+        }
+    }
+
+}
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/ArgCheck.java]: There are too many files with `2 spaces` as indent

File: android/src/main/java/com/gettipsi/stripe/util/Converters.java [hunk 25]
```diff
@@ -0,0 +1,421 @@
...
+
+    return newToken;
+  }
+
+  public static WritableMap putExtraToTokenMap(final WritableMap tokenMap, UserAddress billingAddress, UserAddress shippingAddress) {
+    ArgCheck.nonNull(tokenMap);
+
+    WritableMap extra = Arguments.createMap();
+
+    if(billingAddress != null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Converters.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/Converters.java [hunk 26]
```diff
@@ -0,0 +1,421 @@
...
+      extra.putMap("billingContact", convertAddressToWritableMap(billingAddress));
+    }
+
+    if(shippingAddress != null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Converters.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/Converters.java [hunk 27]
```diff
@@ -0,0 +1,421 @@
...
+
+    tokenMap.putMap("extra", extra);
+
+    return tokenMap;
+  }
+
+  private static WritableMap convertCardToWritableMap(final Card card) {
+    WritableMap result = Arguments.createMap();
+
+    if(card == null) return result;
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Converters.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/Converters.java [hunk 28]
```diff
@@ -0,0 +1,421 @@
...
+    result.putString("country", card.getCountry() );
+    result.putString("currency", card.getCurrency() );
+
+    return result;
+  }
+
+  public static WritableMap convertBankAccountToWritableMap(BankAccount account) {
+    WritableMap result = Arguments.createMap();
+
+    if(account == null) return result;
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Converters.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/Converters.java [hunk 29]
```diff
@@ -0,0 +1,421 @@
...
+
+  public static String getValue(final ReadableMap map, final String key) {
+    return getValue(map, key, (String) null);
+  }
+
+  public static Collection<String> getAllowedShippingCountryCodes(final ReadableMap map) {
+    ArrayList<String> allowedCountryCodesForShipping = new ArrayList<>();
+    ReadableArray countries = getValue(map, "shipping_countries", (ReadableArray) null);
+
+    if(countries != null){
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Converters.java]:
```java
if (countries != null) {
```

File: android/src/main/java/com/gettipsi/stripe/util/Converters.java [hunk 30]
```diff
@@ -0,0 +1,421 @@
...
+    }
+
+    return allowedCountryCodesForShipping;
+  }
+
+  public static ArrayList<CountrySpecification> getAllowedShippingCountries(final ReadableMap map) {
+    ArrayList<CountrySpecification> allowedCountriesForShipping = new ArrayList<>();
+    ReadableArray countries = getValue(map, "shipping_countries", (ReadableArray) null);
+
+    if(countries != null){
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Converters.java]:
```java
if (countries != null) {
```

File: android/src/main/java/com/gettipsi/stripe/util/Converters.java [hunk 31]
```diff
@@ -0,0 +1,421 @@
...
+      map.putArray(key, (WritableNativeArray) object);
+    } else {
+
+    }
+  }
+
+  public static WritableMap convertAddressToWritableMap(final UserAddress address){
+    WritableMap result = Arguments.createMap();
+
+    if(address == null) return result;
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Converters.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/Converters.java [hunk 32]
```diff
@@ -0,0 +1,421 @@
...
+  }
+
+  public static void putIfNotEmpty(final WritableMap map, final String key, final String value) {
+    if (!TextUtils.isEmpty(value)) {
+      map.putString(key, value);
+    }
+  }
+
+  public static UserAddress getBillingAddress(PaymentData paymentData) {
+    if(paymentData != null && paymentData.getCardInfo() != null) {
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Converters.java]:
```java
if (
```

File: android/src/main/java/com/gettipsi/stripe/util/Fun0.java [hunk 33]
```diff
@@ -0,0 +1,5 @@
+package com.gettipsi.stripe.util;
+
+public interface Fun0<R> {
+    R call();
+}
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/util/Fun0.java]: indent 2 spaces

File: android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java (hunk 8 above)
Comment (paynd) [reply android/src/main/java/com/gettipsi/stripe/GoogleApiPayFlowImpl.java]: @ngoryachev Not at all. It looks forgotten to simplify.

**Comment by cybergrind** [comment]
@ngoryachev @isnifer @paynd thank you guys

**Pull request merged**